### PR TITLE
Sync ownership state with blockchain (Marketplace contract)

### LINKED ISSUE #90: Marketplace contract does not assert if one is an admin for an asset
Parent issue: https://github.com/eigerco/nebula/issues/81

The code in reference:
https://github.com/eigerco/nebula/blob/e431bb39e56e66062cd2db24775d041b0b6fc479/contracts/marketplace/src/lib.rs#L78-L81

### PR DESCRIPTION
**Pull request by eloylp**
**Referenced issue**

Closes #90

**Summary of the main changes**

- As the title says, sync internal state of the contract regarding ownership with the blockchain.
- Make the contract the asset owner while the assets are listed.
- Update public API.
- Add more test coverage. (auths and events mainly)
- Add a helper script for simulating NFTs during integration tests.
- Update and fix integration tests.

**Reviewer recommendations**

- Review each commit. One can do this incrementally. 
- Ensure all the corner cases are covered in tests for the fix.
- Ensure the new additions makes sense in both , production code and tests.
- Ensure CI is passing.

### CODE REVIEW

**Comment by mariopil** [comment]
Integration tests are failing now, I guess the problem is with asset
```
[cargo-make] INFO - Running Task: invoke_create_listing
2023-10-24T06:24:22.726528Z ERROR soroban_cli::log::diagnostic_event: 0: "AAAAAAAAAAAAAAAAAAAAAgAAAAAAAAADAAAADwAAAAdmbl9jYWxsAAAAAA0AAAAgG4br9Ie/BXSeLznrbpSmOofRBVxMNQgjyXXUaKdSdJkAAAAPAAAADmNyZWF0ZV9saXN0aW5nAAAAAAAQAAAAAQAAAAMAAAASAAAAAAAAAAC5vSmUIXhYcmRviN6+76fN51wvyxPb2NK7jlPk5i+R0QAAABIAAAABfcHs35M1GZ/JkY2+DHMs4dEUaqjynMnDYK/Gp0eulN8AAAAKAAAAAAAAAAAAAAAAAAAAZA=="
2023-10-24T06:24:22.726626Z ERROR soroban_cli::log::diagnostic_event: 1: "AAAAAAAAAAAAAAABG4br9Ie/BXSeLznrbpSmOofRBVxMNQgjyXXUaKdSdJkAAAACAAAAAAAAAAMAAAAPAAAAB2ZuX2NhbGwAAAAADQAAACB9wezfkzUZn8mRjb4Mcyzh0RRqqPKcycNgr8anR66U3wAAAA8AAAAJc2V0X2FkbWluAAAAAAAAEgAAAAEbhuv0h78FdJ4vOetulKY6h9EFXEw1CCPJddRop1J0mQ=="
2023-10-24T06:24:22.726648Z ERROR soroban_cli::log::diagnostic_event: 2: "AAAAAAAAAAAAAAABG4br9Ie/BXSeLznrbpSmOofRBVxMNQgjyXXUaKdSdJkAAAACAAAAAAAAAAIAAAAPAAAABWVycm9yAAAAAAAAAgAAAAIAAAAHAAAAEAAAAAEAAAADAAAADgAAABRjb250cmFjdCBjYWxsIGZhaWxlZAAAAA8AAAAJc2V0X2FkbWluAAAAAAAAEAAAAAEAAAABAAAAEgAAAAEbhuv0h78FdJ4vOetulKY6h9EFXEw1CCPJddRop1J0mQ=="
2023-10-24T06:24:22.726703Z ERROR soroban_cli::log::diagnostic_event: 3: "AAAAAAAAAAAAAAABG4br9Ie/BXSeLznrbpSmOofRBVxMNQgjyXXUaKdSdJkAAAACAAAAAAAAAAIAAAAPAAAABWVycm9yAAAAAAAAAgAAAAIAAAAHAAAADgAAAEBlc2NhbGF0aW5nIGVycm9yIHRvIFZNIHRyYXAgZnJvbSBmYWlsZWQgaG9zdCBmdW5jdGlvbiBjYWxsOiBjYWxs"
2023-10-24T06:24:22.726732Z ERROR soroban_cli::log::diagnostic_event: 4: "AAAAAAAAAAAAAAABG4br9Ie/BXSeLznrbpSmOofRBVxMNQgjyXXUaKdSdJkAAAACAAAAAAAAAAEAAAAPAAAAA2xvZwAAAAAQAAAAAQAAAAMAAAAOAAAAHlZNIGNhbGwgdHJhcHBlZCB3aXRoIEhvc3RFcnJvcgAAAAAADwAAAA5jcmVhdGVfbGlzdGluZwAAAAAAAgAAAAIAAAAH"
error: transaction simulation failed: host invocation failed

Caused by:
    HostError: Error(Context, InternalError)
    
    Event log (newest first):
       0: [Diagnostic Event] contract:1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499, topics:[error, Error(Context, InternalError)], data:"escalating error to VM trap from failed host function call: call"
       1: [Diagnostic Event] contract:1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499, topics:[error, Error(Context, InternalError)], data:["contract call failed", set_admin, [Address(Contract(1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499))]]
       2: [Diagnostic Event] contract:1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499, topics:[fn_call, Bytes(7dc1ecdf9335199fc9918dbe0c732ce1d1146aa8f29cc9c360afc6a747ae94df), set_admin], data:Address(Contract(1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499))
       3: [Diagnostic Event] topics:[fn_call, Bytes(1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499), create_listing], data:[Address(Account(b9bd299421785872646f88debeefa7cde75c2fcb13dbd8d2bb8e53e4e62f91d1)), Address(Contract(7dc1ecdf9335199fc9918dbe0c732ce1d1146aa8f29cc9c360afc6a747ae94df)), 100]
    
    Backtrace (newest first):
       0: soroban_wasmi::engine::EngineExecutor::execute_wasm_func
       1: soroban_wasmi::engine::EngineInner::execute_func
       2: soroban_wasmi::func::Func::call
       3: soroban_env_host::vm::Vm::invoke_function_raw
       4: soroban_env_host::host::frame::<impl soroban_env_host::host::Host>::with_frame
       5: soroban_env_host::host::frame::<impl soroban_env_host::host::Host>::call_n_internal
       6: soroban_env_host::host::frame::<impl soroban_env_host::host::Host>::invoke_function
       7: preflight::preflight::preflight_invoke_hf_op
       8: preflight::preflight_invoke_hf_op::{{closure}}
       9: core::ops::function::FnOnce::call_once{{vtable.shim}}
      10: preflight::catch_preflight_panic
      11: _cgo_0b49d6ed4a0b_Cfunc_preflight_invoke_hf_op
                 at tmp/go-build/cgo-gcc-prolog:103:11
      12: runtime.asmcgocall
```

**Comment by eloylp** [comment]
> Integration tests are failing now, I guess the problem is with asset
> 
> ```
> [cargo-make] INFO - Running Task: invoke_create_listing
> 2023-10-24T06:24:22.726528Z ERROR soroban_cli::log::diagnostic_event: 0: "AAAAAAAAAAAAAAAAAAAAAgAAAAAAAAADAAAADwAAAAdmbl9jYWxsAAAAAA0AAAAgG4br9Ie/BXSeLznrbpSmOofRBVxMNQgjyXXUaKdSdJkAAAAPAAAADmNyZWF0ZV9saXN0aW5nAAAAAAAQAAAAAQAAAAMAAAASAAAAAAAAAAC5vSmUIXhYcmRviN6+76fN51wvyxPb2NK7jlPk5i+R0QAAABIAAAABfcHs35M1GZ/JkY2+DHMs4dEUaqjynMnDYK/Gp0eulN8AAAAKAAAAAAAAAAAAAAAAAAAAZA=="
> 2023-10-24T06:24:22.726626Z ERROR soroban_cli::log::diagnostic_event: 1: "AAAAAAAAAAAAAAABG4br9Ie/BXSeLznrbpSmOofRBVxMNQgjyXXUaKdSdJkAAAACAAAAAAAAAAMAAAAPAAAAB2ZuX2NhbGwAAAAADQAAACB9wezfkzUZn8mRjb4Mcyzh0RRqqPKcycNgr8anR66U3wAAAA8AAAAJc2V0X2FkbWluAAAAAAAAEgAAAAEbhuv0h78FdJ4vOetulKY6h9EFXEw1CCPJddRop1J0mQ=="
> 2023-10-24T06:24:22.726648Z ERROR soroban_cli::log::diagnostic_event: 2: "AAAAAAAAAAAAAAABG4br9Ie/BXSeLznrbpSmOofRBVxMNQgjyXXUaKdSdJkAAAACAAAAAAAAAAIAAAAPAAAABWVycm9yAAAAAAAAAgAAAAIAAAAHAAAAEAAAAAEAAAADAAAADgAAABRjb250cmFjdCBjYWxsIGZhaWxlZAAAAA8AAAAJc2V0X2FkbWluAAAAAAAAEAAAAAEAAAABAAAAEgAAAAEbhuv0h78FdJ4vOetulKY6h9EFXEw1CCPJddRop1J0mQ=="
> 2023-10-24T06:24:22.726703Z ERROR soroban_cli::log::diagnostic_event: 3: "AAAAAAAAAAAAAAABG4br9Ie/BXSeLznrbpSmOofRBVxMNQgjyXXUaKdSdJkAAAACAAAAAAAAAAIAAAAPAAAABWVycm9yAAAAAAAAAgAAAAIAAAAHAAAADgAAAEBlc2NhbGF0aW5nIGVycm9yIHRvIFZNIHRyYXAgZnJvbSBmYWlsZWQgaG9zdCBmdW5jdGlvbiBjYWxsOiBjYWxs"
> 2023-10-24T06:24:22.726732Z ERROR soroban_cli::log::diagnostic_event: 4: "AAAAAAAAAAAAAAABG4br9Ie/BXSeLznrbpSmOofRBVxMNQgjyXXUaKdSdJkAAAACAAAAAAAAAAEAAAAPAAAAA2xvZwAAAAAQAAAAAQAAAAMAAAAOAAAAHlZNIGNhbGwgdHJhcHBlZCB3aXRoIEhvc3RFcnJvcgAAAAAADwAAAA5jcmVhdGVfbGlzdGluZwAAAAAAAgAAAAIAAAAH"
> error: transaction simulation failed: host invocation failed
> 
> Caused by:
>     HostError: Error(Context, InternalError)
>     
>     Event log (newest first):
>        0: [Diagnostic Event] contract:1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499, topics:[error, Error(Context, InternalError)], data:"escalating error to VM trap from failed host function call: call"
>        1: [Diagnostic Event] contract:1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499, topics:[error, Error(Context, InternalError)], data:["contract call failed", set_admin, [Address(Contract(1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499))]]
>        2: [Diagnostic Event] contract:1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499, topics:[fn_call, Bytes(7dc1ecdf9335199fc9918dbe0c732ce1d1146aa8f29cc9c360afc6a747ae94df), set_admin], data:Address(Contract(1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499))
>        3: [Diagnostic Event] topics:[fn_call, Bytes(1b86ebf487bf05749e2f39eb6e94a63a87d1055c4c350823c975d468a7527499), create_listing], data:[Address(Account(b9bd299421785872646f88debeefa7cde75c2fcb13dbd8d2bb8e53e4e62f91d1)), Address(Contract(7dc1ecdf9335199fc9918dbe0c732ce1d1146aa8f29cc9c360afc6a747ae94df)), 100]
>     
>     Backtrace (newest first):
>        0: soroban_wasmi::engine::EngineExecutor::execute_wasm_func
>        1: soroban_wasmi::engine::EngineInner::execute_func
>        2: soroban_wasmi::func::Func::call
>        3: soroban_env_host::vm::Vm::invoke_function_raw
>        4: soroban_env_host::host::frame::<impl soroban_env_host::host::Host>::with_frame
>        5: soroban_env_host::host::frame::<impl soroban_env_host::host::Host>::call_n_internal
>        6: soroban_env_host::host::frame::<impl soroban_env_host::host::Host>::invoke_function
>        7: preflight::preflight::preflight_invoke_hf_op
>        8: preflight::preflight_invoke_hf_op::{{closure}}
>        9: core::ops::function::FnOnce::call_once{{vtable.shim}}
>       10: preflight::catch_preflight_panic
>       11: _cgo_0b49d6ed4a0b_Cfunc_preflight_invoke_hf_op
>                  at tmp/go-build/cgo-gcc-prolog:103:11
>       12: runtime.asmcgocall
> ```

Yes. I always miss integration tests are there :sweat: . Its nice they are there :smile: . Looks like current integration tests are trying to list in the marketplace the stellar native asset, from the `Makefile.toml`:

```toml
[tasks.get_asset]
plugin = "retry_script"
script = '''
soroban lab token id --asset native --network ${network} > /tmp/asset_id.txt
'''
```

Then `/tmp/asset_id.txt` is used at create listing:

```toml
[tasks.invoke_create_listing]
plugin = "retry_script"
script = '''
    soroban contract invoke \
    --id $(cat /tmp/contract_id.txt) \
    --source trader_1 \
    --network ${network} \
    -- \
    create_listing \
    --seller $(soroban config identity address trader_1) \
    --asset $(cat /tmp/asset_id.txt) \
    --price 100 \
'''
```
I think we should create an stellar asset in a previous test script, and use it for tests. I will investigate how to do that.

**Review by geofmureithi — CHANGES_REQUESTED** [review]
Updates are not provided to documentation in the md and in rust docs, please update those.

File: contracts/marketplace/Makefile.toml [hunk 1]
```diff
@@ -53,16 +48,13 @@ script = '''
...
-soroban lab token id --asset native --network ${network} > /tmp/token_id.txt
-'''
-
-[tasks.get_asset]
-plugin = "retry_script"
-script = '''
-soroban lab token id --asset native --network ${network} > /tmp/asset_id.txt
+  soroban config identity generate asset_issuer_account
+  curl "${friendbot}?addr=$(soroban config identity address asset_issuer_account)"
+  ./asset_test_gen/asset_test_gen -i $(soroban config identity show asset_issuer_account) -d $(soroban config identity show trader_1) -r $(soroban config identity show trader_2),$(soroban config identity show trader_3) > /tmp/asset_address_pair.txt
```
[the remark this answers is not in the document]
Comment (eloylp) [reply contracts/marketplace/Makefile.toml:56]: I was following this [docs](https://developers.stellar.org/docs/issuing-assets/how-to-issue-an-asset#full-code-sample), and there is no Rust example there. I went for the Go one because its the only one compiled language there. Would love to have that done in Rust, but for the size/time available, Go was a direct win.

File: contracts/marketplace/asset_test_gen/asset_test_gen
[the remark this answers is not in the document]
Comment (eloylp) [reply contracts/marketplace/asset_test_gen/asset_test_gen:1]: Made a README.md here :) - https://github.com/eigerco/nebula/pull/96/files/eccf1ddd05805ff1df9214a7224b765bccf6569d#diff-675bf3160e53b1cfda6a156052678bafff2744e6d005e687856372a03f020b6a

File: contracts/marketplace/asset_test_gen/go.mod
[the remark this answers is not in the document]
Comment (eloylp) [reply contracts/marketplace/asset_test_gen/go.mod:1]: Yes, as they hold dependencies versions and ensure deps checksums.

**Comment by geofmureithi** [comment]
@eloylp  I hope I did not break anything, I accidentally pushed but reverted. I created a PR to illustrate my suggestion.

**Comment by eloylp** [comment]
> Updates are not provided to documentation in the md and in rust docs, please update those.

This was fixed at:

* 12628cb2d9dc2b13c3a080cadc2626d58fe0c09f
* 6168a705bed5715bd98252c2f018cbdb8ee7e1d8
* 377c1e3b9d9862646dd8b97e5548294b6d673270

**Comment by eloylp** [comment]
@geofmureithi as we commented, added the clarification regarding trustlines creation in docs and referenced the issue (#103 ) on them . I think we can move this forward.

* https://github.com/eigerco/nebula/pull/96/commits/e50c43ebc90f0a8e7f6321482164332d158da066

File: contracts/marketplace/asset_test_gen/asset_test_gen
[the remark this answers is not in the document]
Comment (geofmureithi) [reply contracts/marketplace/asset_test_gen/asset_test_gen:1]: Sorted by #101

**Review by geofmureithi — APPROVED** [review]
LGTM, let me get the asset creation part working.

File: contracts/marketplace/asset_test_gen/main.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	request := horizonclient.AccountRequest{AccountID: issuerKeyPair.Address()}
+	issuerAccount, err := client.AccountDetail(request)
+	mustBeOK(err)
+
+	// Create an object to represent the new asset
+	nft_code := "EigerNFT"
+	nft := txnbuild.CreditAsset{Code: nft_code, Issuer: issuerKeyPair.Address()}
+
+	threshold := txnbuild.Threshold(0)
+	tx_ops := []txnbuild.Operation{
```
[the remark this answers is not in the document]
Comment (eloylp) [reply contracts/marketplace/asset_test_gen/main.go:50]: Do we have a Rust library for interacting this way ? Where ? :raccoon: :)

**Comment by eloylp** [comment]
Hell, finally, after 3 CI restarts it went through :sweat: . @mariopil could you review/approve ?
